bundles/org.openhab.binding.tibber/src/main/java/org/openhab/binding/tibber/internal/handler/TibberHandler.java

```diff
@@ -0,0 +1,471 @@
+        if (pollingJob == null) {
+            pollingJob = scheduler.scheduleWithFixedDelay(() -> {
+                try {
+                    updateRequest();
+                } catch (IOException e) {
+                    logger.warn("IO Exception: ", e);
+                }
+            }, 1, refresh, TimeUnit.MINUTES);
+        }
+    }
+
+    public void updateRequest() throws IOException {
+        getURLInput(BASE_URL);
+
+        if ("true".equals(rtEnabled) && !isConnected()) {
+            try {
+                open();
+            } catch (Exception e) {
+                logger.warn("Exception: ", e);
+            }
+        }
+    }
+
+    public void updateChannel(String channelID, String channelValue) {
+        if (!channelValue.contains("null")) {
+            if (channelID.contains("consumption") || channelID.contains("Consumption")
+                    || channelID.contains("accumulatedProduction")) {
+                updateState(channelID, new QuantityType<>(new BigDecimal(channelValue), SmartHomeUnits.KILOWATT_HOUR));
+            } else if (channelID.contains("power") || channelID.contains("Power")) {
+                updateState(channelID, new QuantityType<>(new BigDecimal(channelValue), SmartHomeUnits.WATT));
+            } else if (channelID.contains("voltage")) {
+                updateState(channelID, new QuantityType<>(new BigDecimal(channelValue), SmartHomeUnits.VOLT));
+            } else if (channelID.contains("live_current")) {
+                updateState(channelID, new QuantityType<>(new BigDecimal(channelValue), SmartHomeUnits.AMPERE));
+            } else {
+                updateState(channelID, new DecimalType(channelValue));
+            }
+        }
+    }
+
+    public void thingStatusChanged(ThingStatusInfo thingStatusInfo) {
+        logger.debug("Thing Status updated to {} for device: {}", thingStatusInfo.getStatus(), getThing().getUID());
+        if (thingStatusInfo.getStatus() != ThingStatus.ONLINE) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                    "Unable to communicate with Tibber API");
+        }
+    }
+
+    @Override
+    public void dispose() {
+        ScheduledFuture<?> pollingJob = this.pollingJob;
+        if (pollingJob != null) {
+            pollingJob.cancel(true);
+            this.pollingJob = null;
+        }
+        if (isConnected()) {
+            close();
+        }
+        super.dispose();
+    }
+
+    public void open() throws Exception {
+        if (isConnected()) {
+            logger.debug("Open: connection is already open");
+        } else {
+            logger.info("Connecting to: {}", SUBSCRIPTION_URL);
+            TibberConfiguration tibberConfig = getConfigAs(TibberConfiguration.class);
+
+            sslContextFactory.setTrustAll(true);
+            sslContextFactory.setEndpointIdentificationAlgorithm(null);
+
+            WebSocketClient newClient = new WebSocketClient(sslContextFactory);
```

Comment:
The lifecycle of this client should be tied to the handler's lifecycle. You should start and stop it this handler's `initialize()` and `dispose()` methods respectively. Also you should specify the executor that this WebSocketClient should use so that you can make sure that the names of the threads can be traced back to this binding.

Using `private final Executor websocketExecutor = ThreadPoolManager.getPool("tibber.websocket");` would suffice.

Reply:
Unsure what is requested here: `initialize()` and `dispose()` triggers `open()` and `close` respectively. `open()` starts and connects the client, whereas `close()` will close and nullify the session, which again should stop the client.

Reply:
What I'm pointing out is that `WebSocketClient` is a class that has a lifecycle, and as such should be managed. You need to make `WebSocketClient` a field and start and stop the client in your handler's initialize and dispose methods.

Reply:
Ok, have rewritten code, and have implemented `Executor websocketExecutor` as proposed.

Although, as there is a prerequisite to be met before a websocket connection is to be established, I am uncertain if the `WebSocketClient` should `start()` and `stop()` as part of `initialize()` and `dispose()`. I have therefore made the `WebSocketClient` indirectly dependent to `initialize()` and `dispose()`:

**1. Start**
Initialize() --> Method which checks prerequisite --> WebSocket open() --> WebSocketClient.start / WebSocketClient.connect

**2. Stop**
Dispose --> Check if WebSocket is connected --> WebSocket close() --> WebSocketClient.stop() (and nullify client)

Reply:
Please call start/stop inside initialize/dispose

Reply:
With my little experience with OH binding development, could I ask why it is not "good enough" to have start indirectly initialised, meaning as part of directly initialised method? The websocket part is "optional", and does not influence the Thing status. 

The thing is set up (based on vendor spesicic API), to respond to http request by default, and if a prerequisite is met (customer option), additional data is retrieved through websocket. For me it seems unnecessary to initialise and start a client if not required?

For dispose / close it is handled, i.e. checking if null, and stop/nullify if not (to follow in next commit).

Reply:
Ok, I keep forgetting that websocket is optional. In that case make the WebSocketClient a nullable field and just make sure to call `stop()` on the client (if it is not null) inside the handler's `dispose()`. You can start the websocket at any time you feel like. I just need to make sure that when the handler is disposed, the websocket client (if used) is disposed as well.

Reply:
`WebSocketClient` now stopped and nullified (if not null) as part of `dispose()`.